stew/io2.nim

```diff
@@ -0,0 +1,1308 @@
+
+proc removeFile*(path: string): IoResult[void] =
+  ## Deletes a file ``path``.
+  ##
+  ## Procedure will not fail, if file do not exist.
+  when defined(posix):
+    if posix.unlink(path) != 0'i32:
+      let errCode = ioLastError()
+      if errCode == ENOENT:
+        ok()
+      else:
+        err(errCode)
+    else:
+      ok()
+  elif defined(windows):
+    if deleteFileW(newWideCString(path)) == 0:
+      let errCode = ioLastError()
+      if errCode == ERROR_FILE_NOT_FOUND:
+        ok()
+      else:
+        err(errCode)
+    else:
+      ok()
+
+proc isFile*(path: string): bool =
+  ## Returns ``true`` if ``path`` exists and is a regular file or symlink.
+  when defined(posix):
+    var a: posix.Stat
+    let res = posix.stat(path, a)
+    if res == -1:
+      false
+    else:
+      posix.S_ISREG(a.st_mode)
+  elif defined(windows):
+    let res = getFileAttributes(newWideCString(path))
+    if res == INVALID_FILE_ATTRIBUTES:
+      false
+    else:
+      (res and FILE_ATTRIBUTE_DIRECTORY) == 0'u32
+
+proc isDir*(path: string): bool =
+  ## Returns ``true`` if ``path`` exists and is a directory.
+  when defined(posix):
+    var a: posix.Stat
+    let res = posix.stat(path, a)
+    if res == -1:
+      false
+    else:
+      posix.S_ISDIR(a.st_mode)
+  elif defined(windows):
+    let res = getFileAttributes(newWideCString(path))
+    if res == INVALID_FILE_ATTRIBUTES:
+      false
+    else:
+      (res and FILE_ATTRIBUTE_DIRECTORY) == FILE_ATTRIBUTE_DIRECTORY
+
+proc getPathItems(path: string, reverse: bool): seq[string] =
+  var paths: seq[string]
+  let root = $DirSep
+
+  when defined(windows):
+    let (drive, dpath) = splitDrive(path)
+    var curpath = dpath
+  else:
+    var curpath = path
+
+  normPathEnd(curpath, trailingSep = false)
+  while true:
+    let curbase = basename(curpath)
+    let curdir = dirname(curpath)
+    curpath = curdir
+    if len(curbase) > 0:
+      when defined(posix):
+        if len(curdir) > 0 and curdir != root:
+          paths.add(curdir & DirSep & curbase)
+        else:
+          paths.add(curdir & curbase)
+      elif defined(windows):
+        if len(curdir) > 0 and curdir != root:
+          paths.add(drive & curdir & DirSep & curbase)
+        else:
+          paths.add(drive & curdir & curbase)
+    else:
+      break
+  if reverse:
+    paths.reverse()
+  paths
+
+proc createPath*(path: string, createMode: int = 0o755): IoResult[void] =
+  ## Creates the full path ``path`` with mode ``createMode``.
+  ##
+  ## Path may contain several subfolders that do not exist yet.
+  ## The full path is created. If this fails, error will be returned.
+  ##
+  ## It does **not** fail if the folder already exists because for
+  ## most usages this does not indicate an error.
+  let paths = getPathItems(path, true)
+  for item in paths:
+    let res = rawCreateDir(item, createMode)
+    if res.isErr():
+      return err(res.error)
+  ok()
+
+proc toSet*(mask: int): Permissions =
```

Comment:
This is a terrible name for a proc that acts on a primitive type like `int`. Maybe `toPermissions` would work better?